GillianCore/engine/FOLogic/Reduction.ml

```diff
@@ -1714,393 +1359,965 @@ and reduce_lexpr_loop
-            | exn -> raise exn
-          in
-          (flel, fler)
-        in
-        let def = Expr.BinOp (flel, op, fler) in
-        match (flel, fler) with
-        | Lit ll, Lit lr -> (
-            try
-              Lit
-                (CExprEval.evaluate_binop (CStore.init []) op (Lit ll) (Lit lr))
-            with
+    (* -------------------------
+                 UnOp
+       ------------------------- *)
+    | UnOp (NumToInt, UnOp (IntToNum, le)) -> f le
+    | UnOp (IntToNum, UnOp (NumToInt, le)) when PFS.mem pfs (UnOp (IsInt, le))
+      -> f le
+    (* Number-to-string-to-number-to-string-to... *)
+    | UnOp (ToNumberOp, UnOp (ToStringOp, le)) -> (
+        let fle = f le in
+        match fle with
+        | Lit (Num _) -> fle
+        | _ -> (
+            let tfle, how = Typing.type_lexpr gamma fle in
+            match (how, tfle) with
+            | true, Some NumberType -> fle
+            | _, _ -> UnOp (ToNumberOp, UnOp (ToStringOp, fle))))
+    | UnOp (LstRev, UnOp (LstRev, le)) -> f le
+    (* Less than and lessthaneq *)
+    | UnOp (Not, BinOp (le1, FLessThan, le2)) ->
+        f (BinOp (f le2, FLessThanEqual, f le1))
+    | UnOp (Not, BinOp (le1, FLessThanEqual, le2)) ->
+        f (BinOp (f le2, FLessThan, f le1))
+    | UnOp (Not, BinOp (le1, ILessThan, le2)) ->
+        f (BinOp (f le2, ILessThanEqual, f le1))
+    | UnOp (Not, BinOp (le1, ILessThanEqual, le2)) ->
+        f (BinOp (f le2, ILessThan, f le1))
+    | UnOp (FUnaryMinus, UnOp (FUnaryMinus, e)) -> f e
+    | UnOp (IUnaryMinus, UnOp (IUnaryMinus, e)) -> f e
+    | UnOp (LstLen, BinOp (_, LstRepeat, e)) -> f e
+    | UnOp (LstLen, LstSub (_, _, e)) -> f e
+    | UnOp (op, le) -> (
+        let fle = f le in
+        let def = Expr.UnOp (op, fle) in
+        match (op, fle) with
+        | _, Lit lit -> (
+            try Lit (CExprEval.evaluate_unop op lit) with
             | CExprEval.TypeError err_msg ->
                 raise (ReductionException (def, err_msg))
             | CExprEval.EvaluationError err_msg ->
                 raise (ReductionException (def, err_msg))
             | e -> raise e)
-        | _ -> (
-            match
-              reduce_binop_inttonum_const matching reduce_lvars pfs gamma flel
-                fler op
-            with
-            | Some e -> e
-            | None -> (
-                match op with
-                | Equal -> (
-                    if Expr.equal flel fler then Lit (Bool true)
-                    else if
-                      PFS.exists
-                        (fun e ->
-                          Formula.equal e (Eq (flel, fler))
-                          || Formula.equal e (Eq (fler, flel)))
-                        pfs
-                    then Lit (Bool true)
-                    else if
-                      PFS.mem pfs (Not (Eq (flel, fler)))
-                      || PFS.mem pfs (Not (Eq (fler, flel)))
-                    then Lit (Bool false)
```

Comment:
I had to remove these, as attempting to reduce `a = x` would always reduce to `true`, since that formula is already in the PFS. Maybe this reduction made sense when Expr/Formula were split? Note that a similar reduction for And/Or/Impl will still work -- this is the only case where it doesn't.